File: Sources/mig/Lexer.swift
// Lexer.swift — Hand-written lexer for MIG .defs files

import Foundation

// MARK: - Token type

enum Token: Equatable {
  case skip
  case routine
  case simpleRoutine
  case subsystem
  case kernelUser
  case kernelServer
  case msgOption
  case useSpecialReplyPort
  case consumeOnSendError
  case msgSeqno
  case waitTime
  case sendTime
  case noWaitTime
  case noSendTime
  case errorProc
  case serverPrefix
  case userPrefix
  case serverDemux
  case rcsId

  case import_
  case uImport
  case sImport
  case iImport
  case dImport

  case in_
  case out
  case inOut
  case userImpl
  case serverImpl
  case requestPort
  case replyPort
  case sReplyPort
  case uReplyPort

  case type_
  case array
  case struct_
  case of

  case inTran
  case outTran
  case destructor
  case cType
  case cUserType
  case userTypeLimit
  case onStackLimit
  case cServerType
  case pointerTo
  case pointerToIfNot
  case valueOf

  case cString
  case secToken
  case userSecToken
  case serverSecToken
  case auditToken
  case userAuditToken
  case serverAuditToken
  case serverContextToken

  case colon  // :
  case semi  // ;
  case comma  // ,
  case plus  // +
  case minus  // -
  case star  // *
  case div  // /
  case lParen  // (
  case rParen  // )
  case equal  // =
  case caret  // ^
  case tilde  // ~
  case lAngle  // <
  case rAngle  // >
  case lBrack  // [
  case rBrack  // ]
  case bar  // |

  case number(UInt)
  case symbolicType(innumber: UInt, instr: String, outnumber: UInt, outstr: String, size: UInt)
  case identifier(String)
  case string(String)
  case qString(String)
  case fileName(String)
  case ipcFlag(IPCTypeFlags)

  case eof
  case illegal(Character)
}

// MARK: - Lexer states

enum LexerState {
  case normal
  case stringMode  // after LookString()
  case fileNameMode  // after LookFileName()
  case qStringMode  // after LookQString()
  case skipToEOL  // after # directive
  case done
}

// MARK: - Lexer

class Lexer {
  private let input: String
  private var index: String.Index
  private var state: LexerState = .normal
  private var savedState: LexerState = .normal
  private var lastToken: Token = .eof

  init(input: String) {
    self.input = input
    self.index = input.startIndex
  }

  // MARK: - Lexer state management

  func lookNormal() {
    savedState = state
    state = .normal
  }

  func lookString() {
    savedState = state
    state = .stringMode
  }

  func lookQString() {
    savedState = state
    state = .qStringMode
  }

  func lookFileName() {
    savedState = state
    state = .fileNameMode
  }

  private func restoreState() {
    state = savedState
  }

  // MARK: - Character helpers

  private var isAtEnd: Bool { index >= input.endIndex }

  private func peek() -> Character? {
    guard index < input.endIndex else { return nil }
    return input[index]
  }

  private func advance() -> Character? {
    guard index < input.endIndex else { return nil }
    let c = input[index]
    index = input.index(after: index)
    return c
  }

  private func match(_ expected: Character) -> Bool {
    guard index < input.endIndex, input[index] == expected else { return false }
    index = input.index(after: index)
    return true
  }

  // MARK: - Lexer main

  func nextToken() -> Token {
    if isAtEnd { return .eof }

    switch state {
    case .normal: return lexNormal()
    case .stringMode: return lexString()
    case .fileNameMode: return lexFileName()
    case .qStringMode: return lexQString()
    case .skipToEOL: return lexSkipToEOL()
    case .done: return .eof
    }
  }

  func allTokens() -> [Token] {
    var tokens: [Token] = []
    while true {
      let tok = nextToken()
      tokens.append(tok)
      if case .eof = tok { break }
    }
    return tokens
  }

  // MARK: - Normal mode

  private func lexNormal() -> Token {
    skipWhitespace()

    guard let c = peek() else { return .eof }

    switch c {
    case "\n":
      _ = advance()
      lineno += 1
      return lexNormal()

    case ":":
      _ = advance()
      return .colon
    case ";":
      _ = advance()
      return .semi
    case ",":
      _ = advance()
      return .comma
    case "+":
      _ = advance()
      return .plus
    case "-":
      _ = advance()
      return .minus
    case "*":
      _ = advance()
      return .star
    case "/":
      _ = advance()
      return .div
    case "(":
      _ = advance()
      return .lParen
    case ")":
      _ = advance()
      return .rParen
    case "=":
      _ = advance()
      return .equal
    case "^":
      _ = advance()
      return .caret
    case "~":
      _ = advance()
      return .tilde
    case "<":
      _ = advance()
      return .lAngle
    case ">":
      _ = advance()
      return .rAngle
    case "[":
      _ = advance()
      return .lBrack
    case "]":
      _ = advance()
      return .rBrack
    case "|":
      _ = advance()
      return .bar

    case "#":
      _ = advance()
      handleSharpDirective()
      return lexNormal()

    case "0"..."9":
      return lexNumber()

    case "A"..."Z", "a"..."z", "_":
      return lexIdentifierOrKeyword()

    default:
      _ = advance()
      return .illegal(c)
    }
  }

  // MARK: - Special modes

  private func lexString() -> Token {
    // In String mode, match a sequence of [-/._$A-Za-z0-9]+
    var str = ""
    while let c = peek() {
      if c.isLetter || c.isNumber || c == "-" || c == "/" || c == "." || c == "_" || c == "$" {
        str.append(advance()!)
      } else {
        break
      }
    }
    if str.isEmpty {
      restoreState()
      return lexNormal()
    }
    restoreState()
    return .string(str)
  }

  private func lexFileName() -> Token {
    // Filename is either a quoted string or an angle-bracket string
    skipWhitespace()
    guard let c = peek() else {
      restoreState()
      return .eof
    }

    if c == "\"" {
      _ = advance()
      var str = ""
      while let ch = peek(), ch != "\"" && ch != "\n" {
        str.append(advance()!)
      }
      if match("\"") {}
      restoreState()
      return .fileName("\"" + str + "\"")
    } else if c == "<" {
      _ = advance()
      var str = ""
      while let ch = peek(), ch != ">" && ch != "\n" {
        str.append(advance()!)
      }
      if match(">") {}
      restoreState()
      return .fileName("<" + str + ">")
    }
    restoreState()
    return lexNormal()
  }

  private func lexQString() -> Token {
    skipWhitespace()
    guard let c = peek(), c == "\"" else {
      restoreState()
      return .eof
    }

    _ = advance()
    var str = ""
    while let ch = peek(), ch != "\"" && ch != "\n" {
      str.append(advance()!)
    }
    if match("\"") {}
    restoreState()
    return .qString("\"" + str + "\"")
  }

  private func lexSkipToEOL() -> Token {
    while let c = peek(), c != "\n" {
      _ = advance()
    }
    if match("\n") {
      lineno += 1
    }
    restoreState()
    return lexNormal()
  }

  // MARK: - Numbers

  private func lexNumber() -> Token {
    var str = ""
    while let c = peek(), c.isNumber {
      str.append(advance()!)
    }
    if let n = UInt(str) {
      return .number(n)
    }
    return .illegal(str.first ?? "?")
  }

  // MARK: - Identifiers and keywords

  private static let keywords: [String: Token] = [
    "routine": .routine,
    "simpleroutine": .simpleRoutine,
    "subsystem": .subsystem,
    "kerneluser": .kernelUser,
    "kernelserver": .kernelServer,
    "msgoption": .msgOption,
    "usespecialreplyport": .useSpecialReplyPort,
    "consumeonsenderror": .consumeOnSendError,
    "msgseqno": .msgSeqno,
    "waittime": .waitTime,
    "sendtime": .sendTime,
    "nowaittime": .noWaitTime,
    "nosendtime": .noSendTime,
    "error": .errorProc,
    "serverprefix": .serverPrefix,
    "userprefix": .userPrefix,
    "serverdemux": .serverDemux,
    "rcsid": .rcsId,
    "import": .import_,
    "uimport": .uImport,
    "simport": .sImport,
    "iimport": .iImport,
    "dimport": .dImport,
    "in": .in_,
    "out": .out,
    "inout": .inOut,
    "userimpl": .userImpl,
    "serverimpl": .serverImpl,
    "requestport": .requestPort,
    "replyport": .replyPort,
    "sreplyport": .sReplyPort,
    "ureplyport": .uReplyPort,
    "type": .type_,
    "array": .array,
    "struct": .struct_,
    "of": .of,
    "intran": .inTran,
    "outtran": .outTran,
    "destructor": .destructor,
    "ctype": .cType,
    "cusertype": .cUserType,
    "usertypelimit": .userTypeLimit,
    "onstacklimit": .onStackLimit,
    "cservertype": .cServerType,
    "pointerto": .pointerTo,
    "pointertoifnot": .pointerToIfNot,
    "valueof": .valueOf,
    "c_string": .cString,
    "sectoken": .secToken,
    "serversectoken": .serverSecToken,
    "usersectoken": .userSecToken,
    "audittoken": .auditToken,
    "serveraudittoken": .serverAuditToken,
    "useraudittoken": .userAuditToken,
    "servercontexttoken": .serverContextToken,
    "skip": .skip,
  ]

  private static let ipcFlags: [(String, IPCTypeFlags)] = [
    ("samecount", .sameCount),
    ("retcode", .retCode),
    ("physicalcopy", .physicalCopy),
    ("overwrite", .overwrite),
    ("dealloc", .dealloc),
    ("notdealloc", .notDealloc),
    ("countinout", .countInOut),
    ("polymorphic", []),
    ("auto", .autoFlag),
    ("const", .constFlag),
  ]

  private static let symbolicTypes: [(String, UInt, UInt, UInt, UInt)] = [
    // name, inType, outType, specialOutType, size
    ("mach_msg_type_unstructured", MACH_MSG_TYPE_UNSTRUCTURED, MACH_MSG_TYPE_UNSTRUCTURED, 0, 0),
    ("mach_msg_type_bit", MACH_MSG_TYPE_BIT, MACH_MSG_TYPE_BIT, 0, 1),
    ("mach_msg_type_boolean", MACH_MSG_TYPE_BOOLEAN, MACH_MSG_TYPE_BOOLEAN, 0, 32),
    ("mach_msg_type_integer_8", MACH_MSG_TYPE_INTEGER_8, MACH_MSG_TYPE_INTEGER_8, 0, 8),
    ("mach_msg_type_integer_16", MACH_MSG_TYPE_INTEGER_16, MACH_MSG_TYPE_INTEGER_16, 0, 16),
    ("mach_msg_type_integer_32", MACH_MSG_TYPE_INTEGER_32, MACH_MSG_TYPE_INTEGER_32, 0, 32),
    ("mach_msg_type_integer_64", MACH_MSG_TYPE_INTEGER_64, MACH_MSG_TYPE_INTEGER_64, 0, 64),
    ("mach_msg_type_real_32", MACH_MSG_TYPE_REAL_32, MACH_MSG_TYPE_REAL_32, 0, 32),
    ("mach_msg_type_real_64", MACH_MSG_TYPE_REAL_64, MACH_MSG_TYPE_REAL_64, 0, 64),
    ("mach_msg_type_char", MACH_MSG_TYPE_CHAR, MACH_MSG_TYPE_CHAR, 0, 8),
    ("mach_msg_type_byte", MACH_MSG_TYPE_BYTE, MACH_MSG_TYPE_BYTE, 0, 8),
    (
      "mach_msg_type_move_receive", MACH_MSG_TYPE_MOVE_RECEIVE, MACH_MSG_TYPE_PORT_RECEIVE, 0,
      PortSize
    ),
    ("mach_msg_type_copy_send", MACH_MSG_TYPE_COPY_SEND, MACH_MSG_TYPE_PORT_SEND, 0, PortSize),
    ("mach_msg_type_make_send", MACH_MSG_TYPE_MAKE_SEND, MACH_MSG_TYPE_PORT_SEND, 0, PortSize),
    ("mach_msg_type_move_send", MACH_MSG_TYPE_MOVE_SEND, MACH_MSG_TYPE_PORT_SEND, 0, PortSize),
    (
      "mach_msg_type_make_send_once", MACH_MSG_TYPE_MAKE_SEND_ONCE, MACH_MSG_TYPE_PORT_SEND_ONCE, 0,
      PortSize
    ),
    (
      "mach_msg_type_move_send_once", MACH_MSG_TYPE_MOVE_SEND_ONCE, MACH_MSG_TYPE_PORT_SEND_ONCE, 0,
      PortSize
    ),
    ("mach_msg_type_port_name", MACH_MSG_TYPE_PORT_NAME, MACH_MSG_TYPE_PORT_NAME, 0, PortSize),
    (
      "mach_msg_type_port_receive", MACH_MSG_TYPE_POLYMORPHIC, MACH_MSG_TYPE_PORT_RECEIVE, 0,
      PortSize
    ),
    ("mach_msg_type_port_send", MACH_MSG_TYPE_POLYMORPHIC, MACH_MSG_TYPE_PORT_SEND, 0, PortSize),
    (
      "mach_msg_type_port_send_once", MACH_MSG_TYPE_POLYMORPHIC, MACH_MSG_TYPE_PORT_SEND_ONCE, 0,
      PortSize
    ),
    (
      "mach_msg_type_polymorphic", MACH_MSG_TYPE_POLYMORPHIC, MACH_MSG_TYPE_POLYMORPHIC, 0, PortSize
    ),
  ]

  private func lexIdentifierOrKeyword() -> Token {
    var str = ""
    while let c = peek(), c.isLetter || c.isNumber || c == "_" {
      str.append(advance()!)
    }

    let lower = str.lowercased()

    // Check keywords first
    if let tok = Lexer.keywords[lower] {
      return tok
    }

    // Check symbolic types
    for (name, inType, outType, _, size) in Lexer.symbolicTypes {
      if lower == name.lowercased() {
        let instr = String(inType)
        let outstr = String(outType)
        return .symbolicType(
          innumber: inType, instr: instr, outnumber: outType, outstr: outstr, size: size)
      }
    }

    // Check IPC flags
    for (name, flag) in Lexer.ipcFlags {
      if lower == name.lowercased() {
        if flag.isEmpty {
          // polymorphic has special handling
          return .symbolicType(
            innumber: MACH_MSG_TYPE_POLYMORPHIC, instr: "MACH_MSG_TYPE_POLYMORPHIC",
            outnumber: MACH_MSG_TYPE_POLYMORPHIC, outstr: "MACH_MSG_TYPE_POLYMORPHIC",
            size: PortSize)
        }
        return .ipcFlag(flag)
      }
    }

    return .identifier(str)
  }

  // MARK: - Helper methods

  private func skipWhitespace() {
    while let c = peek() {
      if c == " " || c == "\t" {
        _ = advance()
      } else if c == "/" {
        // Check for comment
        let nextIdx = input.index(after: index)
        if nextIdx < input.endIndex {
          if input[nextIdx] == "*" {
            skipBlockComment()
            continue
          } else if input[nextIdx] == "/" {
            skipLineComment()
            continue
          }
        }
        break
      } else {
        break
      }
    }
  }

  private func skipBlockComment() {
    _ = advance()  // skip /
    _ = advance()  // skip *
    while let c = advance() {
      if c == "*" && match("/") {
        return
      }
      if c == "\n" { lineno += 1 }
    }
  }

  private func skipLineComment() {
    _ = advance()  // skip /
    _ = advance()  // skip /
    while let c = advance() {
      if c == "\n" {
        lineno += 1
        return
      }
    }
  }

  private func handleSharpDirective() {
    skipWhitespace()

    // Read the directive body up to the first number
    var numStr = ""
    while let c = peek(), c.isNumber {
      numStr.append(advance()!)
    }

    if let lineNum = Int(numStr) {
      lineno = lineNum
    }

    // Look for a filename string
    skipWhitespace()
    if match("\"") {
      var fn = ""
      while let c = peek(), c != "\"" && c != "\n" {
        fn.append(advance()!)
      }
      if match("\"") {
        yyinname = fn
      }
    }

    state = .skipToEOL
  }
}

File: migcom/src/lexer.rs
// SPDX-License-Identifier: MIT
// Lexer for .defs files using winnow

// winnow is used for its AsChar trait to drive the hand-written tokeniser.
use winnow::stream::AsChar;

use crate::types::ipc_type_num::*;

/// Byte offset of a token in the original source
pub type Span = std::ops::Range<usize>;

#[derive(Clone, Debug, PartialEq)]
pub enum Token {
    // keywords
    Routine,
    SimpleRoutine,
    Subsystem,
    MsgOption,
    UseSpecialReplyPort,
    ConsumeOnSendError,
    MsgSeqno,
    WaitTime,
    SendTime,
    NoWaitTime,
    NoSendTime,
    In,
    Out,
    UserImpl,
    ServerImpl,
    SecToken,
    ServerSecToken,
    UserSecToken,
    AuditToken,
    ServerAuditToken,
    UserAuditToken,
    ServerContextToken,
    InOut,
    RequestPort,
    ReplyPort,
    UReplyPort,
    SReplyPort,
    Array,
    Of,
    ErrorProc,
    ServerPrefix,
    UserPrefix,
    ServerDemux,
    RCSId,
    Import,
    UImport,
    SImport,
    DImport,
    IImport,
    Type,
    KernelServer,
    KernelUser,
    Skip,
    Struct,
    InTran,
    OutTran,
    Destructor,
    CType,
    CUserType,
    CServerType,
    CString,
    UserTypeLimit,
    OnStackLimit,
    PointerTo,
    PointerToIfNot,
    ValueOf,
    Novouchers,
    #[allow(dead_code)]
    MachMsg2,
    // IPC flags
    SameCount,
    RetCode,
    PhysicalCopy,
    Overwrite,
    Dealloc,
    NotDealloc,
    CountInOut,
    Auto,
    Const,
    #[allow(dead_code)]
    Polymorphic,
    // symbolic mach types
    SymbolicType {
        in_number: u32,
        in_str: String,
        out_number: u32,
        out_str: String,
        size: u32,
    },
    // punctuation
    Colon,
    Semi,
    Comma,
    Plus,
    Minus,
    Star,
    Div,
    LParen,
    RParen,
    Equal,
    Caret,
    Tilde,
    LAngle,
    RAngle,
    LBrack,
    RBrack,
    Bar,
    // literals
    Number(u32),
    Identifier(String),
    #[allow(dead_code)]
    StringLit(String), // bare unquoted string (path component)
    QStringLit(String), // "quoted"
    FileName(String),   // "quoted" or <angle>
    // errors
    LexError,
}

#[derive(Clone, Debug)]
pub struct Spanned {
    pub token: Token,
    pub span: Span,
}

// ---------------------------------------------------------------------------
// Low-level character helpers
// ---------------------------------------------------------------------------

fn is_ident_start(c: u8) -> bool {
    c.is_alpha() || c == b'_'
}

fn is_ident_continue(c: u8) -> bool {
    c.is_alphanum() || c == b'_'
}

#[allow(dead_code)]
fn is_string_char(c: u8) -> bool {
    // bare String token: [-/._$A-Za-z0-9]+
    c.is_alphanum() || matches!(c, b'-' | b'/' | b'.' | b'_' | b'$')
}

// ---------------------------------------------------------------------------
// Keyword / symbolic-type lookup (case-insensitive)
// ---------------------------------------------------------------------------

fn keyword_or_ident(s: &str) -> Token {
    match s.to_lowercase().as_str() {
        "routine" => Token::Routine,
        "simpleroutine" => Token::SimpleRoutine,
        "subsystem" => Token::Subsystem,
        "msgoption" => Token::MsgOption,
        "usespecialreplyport" => Token::UseSpecialReplyPort,
        "consumeonsenderror" => Token::ConsumeOnSendError,
        "msgseqno" => Token::MsgSeqno,
        "waittime" => Token::WaitTime,
        "sendtime" => Token::SendTime,
        "nowaittime" => Token::NoWaitTime,
        "nosendtime" => Token::NoSendTime,
        "in" => Token::In,
        "out" => Token::Out,
        "userimpl" => Token::UserImpl,
        "serverimpl" => Token::ServerImpl,
        "sectoken" => Token::SecToken,
        "serversectoken" => Token::ServerSecToken,
        "usersectoken" => Token::UserSecToken,
        "audittoken" => Token::AuditToken,
        "serveraudittoken" => Token::ServerAuditToken,
        "useraudittoken" => Token::UserAuditToken,
        "servercontexttoken" => Token::ServerContextToken,
        "inout" => Token::InOut,
        "requestport" => Token::RequestPort,
        "replyport" => Token::ReplyPort,
        "ureplyport" => Token::UReplyPort,
        "sreplyport" => Token::SReplyPort,
        "array" => Token::Array,
        "of" => Token::Of,
        "error" => Token::ErrorProc,
        "serverprefix" => Token::ServerPrefix,
        "userprefix" => Token::UserPrefix,
        "serverdemux" => Token::ServerDemux,
        "rcsid" => Token::RCSId,
        "import" => Token::Import,
        "uimport" => Token::UImport,
        "simport" => Token::SImport,
        "dimport" => Token::DImport,
        "iimport" => Token::IImport,
        "type" => Token::Type,
        "kernelserver" => Token::KernelServer,
        "kerneluser" => Token::KernelUser,
        "skip" => Token::Skip,
        "struct" => Token::Struct,
        "intran" => Token::InTran,
        "outtran" => Token::OutTran,
        "destructor" => Token::Destructor,
        "ctype" => Token::CType,
        "cusertype" => Token::CUserType,
        "cservertype" => Token::CServerType,
        "c_string" => Token::CString,
        "usertypelimit" => Token::UserTypeLimit,
        "onstacklimit" => Token::OnStackLimit,
        "novouchers" => Token::Novouchers,
        // IPC flags (case-insensitive in original)
        "samecount" => Token::SameCount,
        "retcode" => Token::RetCode,
        "physicalcopy" => Token::PhysicalCopy,
        "overwrite" => Token::Overwrite,
        "dealloc" => Token::Dealloc,
        "notdealloc" => Token::NotDealloc,
        "countinout" => Token::CountInOut,
        "polymorphic" => symbolic(
            MACH_MSG_TYPE_POLYMORPHIC,
            MACH_MSG_TYPE_POLYMORPHIC,
            PORT_SIZE,
        ),
        "auto" => Token::Auto,
        "const" => Token::Const,
        // case-SENSITIVE
        _ => Token::Identifier(s.to_owned()),
    }
}

fn symbolic(inn: u32, out: u32, size: u32) -> Token {
    Token::SymbolicType {
        in_number: inn,
        in_str: format!("{inn}"),
        out_number: out,
        out_str: format!("{out}"),
        size,
    }
}

fn case_sensitive_keyword(s: &str) -> Option<Token> {
    Some(match s {
        "PointerTo" => Token::PointerTo,
        "PointerToIfNot" => Token::PointerToIfNot,
        "ValueOf" => Token::ValueOf,
        "MACH_MSG_TYPE_UNSTRUCTURED" => {
            symbolic(MACH_MSG_TYPE_UNSTRUCTURED, MACH_MSG_TYPE_UNSTRUCTURED, 0)
        }
        "MACH_MSG_TYPE_BIT" => symbolic(MACH_MSG_TYPE_BIT, MACH_MSG_TYPE_BIT, 1),
        "MACH_MSG_TYPE_BOOLEAN" => symbolic(MACH_MSG_TYPE_BOOLEAN, MACH_MSG_TYPE_BOOLEAN, 32),
        "MACH_MSG_TYPE_INTEGER_8" => symbolic(MACH_MSG_TYPE_INTEGER_8, MACH_MSG_TYPE_INTEGER_8, 8),
        "MACH_MSG_TYPE_INTEGER_16" => {
            symbolic(MACH_MSG_TYPE_INTEGER_16, MACH_MSG_TYPE_INTEGER_16, 16)
        }
        "MACH_MSG_TYPE_INTEGER_32" => {
            symbolic(MACH_MSG_TYPE_INTEGER_32, MACH_MSG_TYPE_INTEGER_32, 32)
        }
        "MACH_MSG_TYPE_INTEGER_64" => {
            symbolic(MACH_MSG_TYPE_INTEGER_64, MACH_MSG_TYPE_INTEGER_64, 64)
        }
        "MACH_MSG_TYPE_REAL_32" => symbolic(MACH_MSG_TYPE_REAL_32, MACH_MSG_TYPE_REAL_32, 32),
        "MACH_MSG_TYPE_REAL_64" => symbolic(MACH_MSG_TYPE_REAL_64, MACH_MSG_TYPE_REAL_64, 64),
        "MACH_MSG_TYPE_CHAR" => symbolic(MACH_MSG_TYPE_CHAR, MACH_MSG_TYPE_CHAR, 8),
        "MACH_MSG_TYPE_BYTE" => symbolic(MACH_MSG_TYPE_BYTE, MACH_MSG_TYPE_BYTE, 8),
        "MACH_MSG_TYPE_MOVE_RECEIVE" => symbolic(
            MACH_MSG_TYPE_MOVE_RECEIVE,
            MACH_MSG_TYPE_PORT_RECEIVE,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_COPY_SEND" => {
            symbolic(MACH_MSG_TYPE_COPY_SEND, MACH_MSG_TYPE_PORT_SEND, PORT_SIZE)
        }
        "MACH_MSG_TYPE_MAKE_SEND" => {
            symbolic(MACH_MSG_TYPE_MAKE_SEND, MACH_MSG_TYPE_PORT_SEND, PORT_SIZE)
        }
        "MACH_MSG_TYPE_MOVE_SEND" => {
            symbolic(MACH_MSG_TYPE_MOVE_SEND, MACH_MSG_TYPE_PORT_SEND, PORT_SIZE)
        }
        "MACH_MSG_TYPE_MAKE_SEND_ONCE" => symbolic(
            MACH_MSG_TYPE_MAKE_SEND_ONCE,
            MACH_MSG_TYPE_PORT_SEND_ONCE,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_MOVE_SEND_ONCE" => symbolic(
            MACH_MSG_TYPE_MOVE_SEND_ONCE,
            MACH_MSG_TYPE_PORT_SEND_ONCE,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_PORT_NAME" => {
            symbolic(MACH_MSG_TYPE_PORT_NAME, MACH_MSG_TYPE_PORT_NAME, PORT_SIZE)
        }
        "MACH_MSG_TYPE_PORT_RECEIVE" => symbolic(
            MACH_MSG_TYPE_POLYMORPHIC,
            MACH_MSG_TYPE_PORT_RECEIVE,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_PORT_SEND" => symbolic(
            MACH_MSG_TYPE_POLYMORPHIC,
            MACH_MSG_TYPE_PORT_SEND,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_PORT_SEND_ONCE" => symbolic(
            MACH_MSG_TYPE_POLYMORPHIC,
            MACH_MSG_TYPE_PORT_SEND_ONCE,
            PORT_SIZE,
        ),
        "MACH_MSG_TYPE_POLYMORPHIC" => symbolic(
            MACH_MSG_TYPE_POLYMORPHIC,
            MACH_MSG_TYPE_POLYMORPHIC,
            PORT_SIZE,
        ),
        _ => return None,
    })
}

// ---------------------------------------------------------------------------
// Main tokeniser
// ---------------------------------------------------------------------------

/// Tokenise a complete pre-processed .defs source into a vec of Spanned tokens.
pub fn tokenise(src: &str) -> Vec<Spanned> {
    let bytes = src.as_bytes();
    let len = bytes.len();
    let mut pos = 0usize;
    let mut tokens = Vec::new();

    macro_rules! push {
        ($start:expr, $tok:expr) => {
            tokens.push(Spanned {
                token: $tok,
                span: $start..pos,
            });
        };
    }

    while pos < len {
        // Skip whitespace
        while pos < len
            && (bytes[pos] == b' '
                || bytes[pos] == b'\t'
                || bytes[pos] == b'\n'
                || bytes[pos] == b'\r')
        {
            pos += 1;
        }
        if pos >= len {
            break;
        }

        let start = pos;
        let ch = bytes[pos];

        // C-style block comment
        if ch == b'/' && pos + 1 < len && bytes[pos + 1] == b'*' {
            pos += 2;
            while pos + 1 < len && !(bytes[pos] == b'*' && bytes[pos + 1] == b'/') {
                pos += 1;
            }
            pos += 2; // consume */
            continue;
        }

        // C++ line comment
        if ch == b'/' && pos + 1 < len && bytes[pos + 1] == b'/' {
            while pos < len && bytes[pos] != b'\n' {
                pos += 1;
            }
            continue;
        }

        // cpp # directives — skip to end of line (pre-processor already ran)
        if ch == b'#' {
            while pos < len && bytes[pos] != b'\n' {
                pos += 1;
            }
            continue;
        }

        // Quoted string "..."
        if ch == b'"' {
            pos += 1;
            while pos < len && bytes[pos] != b'"' && bytes[pos] != b'\n' {
                pos += 1;
            }
            if pos < len {
                pos += 1;
            } // consume closing "
            let content = &src[start + 1..pos - 1];
            push!(start, Token::QStringLit(content.to_owned()));
            continue;
        }

        // Angle-bracket filename <...>
        if ch == b'<' {
            // Could be filename <foo/bar.h> or LAngle; peek ahead
            let mut j = pos + 1;
            let _valid = true;
            while j < len && bytes[j] != b'>' && bytes[j] != b'\n' {
                j += 1;
            }
            if j < len && bytes[j] == b'>' {
                // looks like a filename
                let content = &src[pos + 1..j];
                pos = j + 1;
                push!(start, Token::FileName(format!("<{content}>")));
                continue;
            }
            // Otherwise treat as LAngle below
        }

        // Numbers
        if ch.is_ascii_digit() {
            while pos < len && bytes[pos].is_ascii_digit() {
                pos += 1;
            }
            let n: u32 = src[start..pos].parse().unwrap_or(0);
            push!(start, Token::Number(n));
            continue;
        }

        // Identifiers / keywords
        if is_ident_start(ch) {
            while pos < len && is_ident_continue(bytes[pos]) {
                pos += 1;
            }
            let word = &src[start..pos];
            // case-sensitive MACH_ and Pointer*/ValueOf check first
            let tok = case_sensitive_keyword(word).unwrap_or_else(|| keyword_or_ident(word));
            push!(start, tok);
            continue;
        }

        // Single-char punctuation
        pos += 1;
        let tok = match ch {
            b':' => Token::Colon,
            b';' => Token::Semi,
            b',' => Token::Comma,
            b'+' => Token::Plus,
            b'-' => Token::Minus,
            b'*' => Token::Star,
            b'/' => Token::Div,
            b'(' => Token::LParen,
            b')' => Token::RParen,
            b'=' => Token::Equal,
            b'^' => Token::Caret,
            b'~' => Token::Tilde,
            b'<' => Token::LAngle,
            b'>' => Token::RAngle,
            b'[' => Token::LBrack,
            b']' => Token::RBrack,
            b'|' => Token::Bar,
            _ => Token::LexError,
        };
        push!(start, tok);
    }

    tokens
}
